java/vector/src/main/java/org/apache/arrow/vector/dictionary/StructSubfieldEncoder.java

```diff
@@ -0,0 +1,184 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.dictionary;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.BaseIntVector;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.complex.StructVector;
+import org.apache.arrow.vector.ipc.message.ArrowFieldNode;
+import org.apache.arrow.vector.types.pojo.DictionaryEncoding;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.arrow.vector.util.TransferPair;
+
+/**
+ * Sub fields encoder/decoder for Dictionary encoded {@link StructVector}.
+ * Notes that child vectors within struct vector can either be dictionary encodeable or not.
+ */
+public class StructSubfieldEncoder {
+
+  private final BufferAllocator allocator;
+
+  private final DictionaryProvider.MapDictionaryProvider provider;
+  private final Map<Long, DictionaryHashTable> dictionaryIdToHashTable;
+
+  /**
+   * Construct an instance.
+   */
+  public StructSubfieldEncoder(
+      BufferAllocator allocator,
+      DictionaryProvider.MapDictionaryProvider provider) {
+
+    this.allocator = allocator;
+    this.provider = provider;
+
+    this.dictionaryIdToHashTable = new HashMap<>();
+
+    provider.getDictionaryIds().forEach(id ->
+        dictionaryIdToHashTable.put(id, new DictionaryHashTable(provider.lookup(id).getVector())));
+  }
+
+  private FieldVector getChildVector(StructVector vector, int index) {
+    return vector.getChildrenFromFields().get(index);
+  }
+
+  private StructVector cloneVector(StructVector vector) {
+
+    final FieldType fieldType = vector.getField().getFieldType();
+    StructVector cloned = (StructVector) fieldType.createNewSingleVector(
+        vector.getField().getName(), allocator, /*schemaCallback=*/null);
+
+    final ArrowFieldNode fieldNode = new ArrowFieldNode(vector.getValueCount(), vector.getNullCount());
+    cloned.loadFieldBuffers(fieldNode, vector.getFieldBuffers());
+
+    return cloned;
+  }
+
+  /**
+   * Dictionary encodes subfields for complex vector with a provided dictionary.
+   * The dictionary must contain all values in the sub fields vector.
+   * @param vector vector to encode
+   * @return dictionary encoded vector
+   */
+  public StructVector encode(StructVector vector, Map<Integer, Long> columnToDictionaryId) {
```

Comment:
please add javadoc for columnToDictionary.

Reply:
fixed.